kimchi/src/circuits/expr.rs

```diff
@@ -1741,13 +1736,236 @@ impl<F: FftField, Column: Copy, ChallengeTerm: Copy> Expr<ConstantExpr<F, Challe
     }
 }
 
+/// Compute the `x`th unnormalized lagrange evaluation
+fn xth_unnormalized_lagrange_eval<
+    'a,
+    F: FftField,
+    ChallengeTerm,
+    Column: Copy,
+    Challenge: Index<ChallengeTerm, Output = F>,
+    Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenge>,
+>(
+    l0_1: F,
+    i: i32,
+    res_domain: Domain,
+    env: &Environment,
+    x: u64,
+) -> F {
+    let k = res_domain as u64;
+    let d1 = env.get_domain(Domain::D1);
+    let n = d1.size;
+    // Renormalize negative values to wrap around at domain size
+    let i = if i < 0 {
+        ((i as isize) + (n as isize)) as usize
+    } else {
+        i as usize
+    };
+    let ii = i as u64;
+    assert!(ii < n);
+    let omega = d1.group_gen;
+    let omega_prime = env.get_domain(res_domain).group_gen;
+    let omega_i = omega.pow([ii]);
+    let omega_minus_i = omega.pow([n - ii]);
+    let omega_x = omega_prime.pow([x]);
+
+    let eval: F = {
+        let v = omega_x - omega_i;
+        v.inverse().unwrap_or(F::zero())
```

Comment:
Note that I copy the behaviour of `batch_inversion` here by just returning `(F::zero()).inverse() == 0` by convention.

cc @marcbeunardeau88